tests/HttpClient/Plugin/HybridOauth2Test.php

```diff
@@ -0,0 +1,198 @@
+use Apigee\Edge\Client;
+use Apigee\Edge\ClientInterface;
+use Apigee\Edge\HttpClient\Utility\Builder;
+use Apigee\Edge\Tests\Test\HttpClient\MockHttpClient;
+use Apigee\Edge\Tests\Test\HttpClient\Plugin\InMemoryOauthTokenStorage;
+use Apigee\Edge\Tests\Test\HttpClient\Plugin\MockHybridOauth2;
+use Apigee\Edge\Tests\Test\HttpClient\Utility\TestJournal;
+use GuzzleHttp\Psr7\Response;
+use Http\Client\Exception\TransferException;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Class HybridOauth2Test.
+ *
+ * @group client
+ * @small
+ */
+class HybridOauth2Test extends TestCase
+{
+    private const API_ENDPOINT = 'http://api.example.com/v1';
+
+    /** @var \Apigee\Edge\Tests\Test\HttpClient\MockHttpClient */
+    protected static $httpClient;
+
+    /** @var \Apigee\Edge\HttpClient\Utility\JournalInterface */
+    private $journal;
+
+    /** @var \Apigee\Edge\ClientInterface */
+    private $client;
+
+    /** @var \Apigee\Edge\HttpClient\Plugin\Authentication\OauthTokenStorageInterface */
+    private $token_storage;
+
+    /**
+     * @inheritdoc
+     */
+    public static function setUpBeforeClass(): void
+    {
+        // Use the Mock HTTP Client for all requests.
+        self::$httpClient = new MockHttpClient();
+        parent::setUpBeforeClass();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->token_storage = new InMemoryOauthTokenStorage();
+        $this->journal = new TestJournal();
+        $this->client = $this->buildClient();
+    }
+
+    /**
+     * @expectedException \Apigee\Edge\Exception\HybridOauth2AuthenticationException
+     * @expectedException \Apigee\Edge\Exception\OauthAccessTokenAuthenticationException
+     */
+    public function testAuthServerError(): void
+    {
+        // Auth server is unavailable.
+        static::$httpClient->addException(new TransferException());
+        $this->client->get('');
+    }
+
+    public function testOauthReAuthenticationAfterExpiredAccessToken(): void
+    {
+        $body = [
+            'access_token' => 'access_token',
+            'expires_in' => 60,
+            'token_type' => 'Bearer',
+        ];
+        // Auth server returns a new access token.
+        static::$httpClient->addResponse(new Response(200, ['Content-Type' => 'application/json'], json_encode((object) $body)));
+
+        // Successful response to an authorised API call from API server.
+        static::$httpClient->addResponse(new Response(200));
+
+        // Another successful response to an authorised API call from API server.
+        static::$httpClient->addResponse(new Response(200));
+
+        $body = [
+            'access_token' => 'access_token',
+            'expires_in' => 60,
+            'token_type' => 'Bearer',
+        ];
+        // Auth server returns a new access token.
+        static::$httpClient->addResponse(new Response(200, ['Content-Type' => 'application/json'], json_encode((object) $body)));
```

Comment:
Again, you are testing the same thing. (Or maybe you should move these setup stuff closer to the actual assertions.)

Reply:
I've added more comments that describe what we are testing and how each of those responses are used, hopefully it's clearer: https://github.com/arlina-espinoza/apigee-client-php/blob/0f8e71ed50bb2082c68ed9534e2a44af64ac211a/tests/HttpClient/Plugin/HybridOauth2Test.php#L96-L168